src/main/java/codeu/service/BBCodeParser.java

```diff
@@ -0,0 +1,29 @@
+package codeu.service;
+
+public class BBCodeParser {
+
+  /**
+   * Parses the input, substituting BBCode tags with their equivalent HTML tags
+   * @param input The text to be parsed
+   * @return The parsed text
+   */
+  public static String parse(String input) {
+    StringBuilder builder = new StringBuilder(input);
+
+    // replace [b] with <b>
+    int i = builder.indexOf("[b]");
+    while (i != -1) {
```

Comment:
As written, I believe this is O(n^2).  The indexOf is going to scan the entire string until it finds what it's looking for and will do so for every [b] element.  A mean user could copy paste a large string of [b] and incur a pretty expensive runtime (unlikely).  For a chat app, that's probably not a big deal, but it could be avoided pretty easily by using java's replaceAll string method.

I don't code much in Java these days (Python and typescript is my main thing at work right now), so my first thought was to use a regexp.  But I googled for BBCode java libraries (your suggestion), and found this:

https://stackoverflow.com/questions/849396/java-bbcode-library

Which, among other things, suggested the replaceAll method.

Reply:
I missed replaceAll when looking at String's API. I'll change it because it is indeed more readable. However, I don't think this would be O(n^2) since, for the indexOf inside the while loop, I actually specify after which index it should search (I use `i`), so I don't actually start searching from the beginning of the string every time (I guess in the end it ends up being the same as replaceAll, but uglier).